hmpy/views/switch.py

```diff
@@ -0,0 +1,118 @@
+"""This module defines the SwitchView class."""
+
+
+from hmpy.views.view import View
+from PyQt5.QtGui import QPainter, QImage, QPixmap, QConicalGradient, QPalette
+from PyQt5.QtCore import Qt
+
+
+class SwitchView(View):
+    """The Switch view creates a clickable switch for triggering actions."""
+
+    _SCALE = 100
+
+    def __init__(self, on=False):
+        """Initialize the switch view.
+
+        :param on: the state of the button"""
+        super().__init__()
+        self._state = on
+        self.mousePressEvent = self._mouse_press_event
+        self._on_toggle = None
+
+        self._pixmap_cache = {}
+
+    def paintEvent(self, event):
+        """Draw the Switch graphic
+
+        :param event: The paint event.
+        """
+
+        # Attempt to fetch the cached pixmap for the current state
+        image = self._pixmap_cache.get(self.on, None)
+
+        if image is None:
+            # If a cached pixmap is not set, generate one and cache it
+            image = self._generate_pixmap(event.rect())
+            self._pixmap_cache[self.on] = image
+
+        # Draw image to view
+        paint = QPainter()
+        paint.begin(self)
+        paint.drawPixmap(self.rect(), image)
+        paint.end()
+
+    @property
+    def on(self):
+        return self._state
+
+    @on.setter
+    def on(self, to):
+        self._state = to
```

Comment:
If we are going to expose on as a property was well as toggle I think we probably have to check for toggle here too.  Maybe move the call to _on_toggle here, with a check if the value is being changed, and then just set .on in the toggle function